### PR TITLE
added duplicate primary key error

### PR DESCRIPTION
**Pull request by no0dles**
allows to catch duplicate primary keys
```
document.save()
	.then(function(r) {
		...
    })
    .catch(Errors.DuplicatePrimaryKey, function () {
        ...
    })
    .error(function (err) {
    	...
    });
```

### CODE REVIEW

**Comment by no0dles** [comment]
allows to catch duplicate primary keys
```
document.save()
	.then(function(r) {
		...
    })
    .catch(Errors.DuplicatePrimaryKey, function () {
        ...
    })
    .error(function (err) {
    	...
    });
```

File: lib/document.js [hunk 1]
```diff
@@ -809,6 +809,12 @@ Document.prototype._onSaved = function(result, docToSave, saveAll, savedModel, r
   var self = this;
 
   if (result.first_error != null) {
+    var duplicatePrimaryKeyRegex = new RegExp('^' + new Errors.DuplicatePrimaryKey().name);
```
Comment (neumino) [inline lib/document.js]: Can you move this in errors.js? And add a `Error.create` method that takes a string and return the appropriate error type.
That would let us keep all the logic related to errors in the same places and would make adding new error type easier.
Thanks!

**Comment by no0dles** [comment]
something like this? 

```
module.exports.create = function (message) {
  for(var key in module.exports) {
    if(module.exports[key].prototype instanceof Error) {
      var regex = new RegExp('^' + new module.exports[key]().name);
      if(message.match(regex))
        return new module.exports[key](message);
    }
  }
  return new Error(message);
};
```
I don't know why wrecker gets stuck on the environment setup...

**Comment by neumino** [comment]
No need for a for loop for now. Also keep the regex in a variable, the name may not perfectly match the error message later.

Thanks!

```
DocumentNotFoundRegex = /.../
module.exports.create = function (message) {
  if (message.match(DocumentNotFoundRegex)) {
    return new DocumentNotFoundError(message);
  }
  return new Error(message);
};
```

**Comment by no0dles** [comment]
after a couple of mistakes I think I got what you want

```
module.exports.create = function (error) {
  if(!(error instanceof Error)) {
    return error;
  }

  if(error.message.match(DocumentNotFoundRegex)) {
    return new DocumentNotFound(error.message);
  }

  if(error.message.match(DuplicatePrimaryKeyRegex)) {
    return new DuplicatePrimaryKey(error.message);
  }

  return error;
};
```

File: lib/errors.js [hunk 2]
```diff
@@ -37,3 +39,35 @@ function ValidationError(message) {
...
+    this.message = message;
+};
+DuplicatePrimaryKey.prototype = new Error();
+DuplicatePrimaryKey.prototype.name = "Duplicate primary key";
+var DuplicatePrimaryKeyRegex = new RegExp('^' + DuplicatePrimaryKey.prototype.name);
+module.exports.DuplicatePrimaryKeyRegex = DuplicatePrimaryKeyRegex;
+module.exports.DuplicatePrimaryKey = DuplicatePrimaryKey;
+
+
+module.exports.create = function (error) {
```
Comment (neumino) [inline lib/errors.js]: Can you make the function take a string instead?

File: lib/document.js [hunk 3]
```diff
@@ -809,7 +809,7 @@ Document.prototype._onSaved = function(result, docToSave, saveAll, savedModel, r
   var self = this;
 
   if (result.first_error != null) {
-    return reject(new Error(result.first_error));
+    return reject(Errors.create(new Error(result.first_error)));
```
Comment (neumino) [inline lib/document.js]: return reject(Errors.create(result.first_error))

**Comment by neumino** [comment]
(if you are reading your mail, forget the two comments I made).

This looks good to me. I'll merge it later tonight. Thanks @no0dles!

**Comment by no0dles** [comment]
would it be possible to merge this pull request this week?

**Comment by neumino** [comment]
My bad,I forgot about it, doing it now.

**Comment by neumino** [comment]
Hum, I ran into a few issues with bluebird (though I don't think it's related to the PR itself).
I'll take another look at it tonight.

**Comment by neumino** [comment]
Err nevermind, I accidentally bump bluebird to v3. I'll revert that and release it tonight.

**Pull request merged**